Frameworks/Foundation/NSDecimal.mm

```diff
@@ -0,0 +1,605 @@
+
+    if (res == NSOrderedAscending) {
+        // R > L
+        NSCalculationError error = _subtraction(result, right, left);
+        if (error != NSCalculationNoError) {
+            INVALIDATE_DECIMAL(result);
+            return error;
+        }
+        result->_isNegative = right->_isNegative;
+        result->_exponent = right->_exponent;
+    } else {
+        NSCalculationError error = _subtraction(result, left, right);
+        if (error != NSCalculationNoError) {
+            INVALIDATE_DECIMAL(result);
+            return error;
+        }
+        result->_isNegative = left->_isNegative;
+        result->_exponent = left->_exponent;
+    }
+
+    NSDecimalCompact(result);
+    return NSCalculationNoError;
+}
+
+/**
+ @Status Caveat
+ @Notes roundingMode is not supported.
+*/
+NSCalculationError NSDecimalAdd(NSDecimal* result,
+                                const NSDecimal* leftOperand,
+                                const NSDecimal* rightOperand,
+                                NSRoundingMode roundingMode) {
+    THROW_NS_IF_FALSE(E_INVALIDARG, (leftOperand != NULL) && (rightOperand != NULL) && (result != NULL));
+    VALIDATE_ARITHMETIC_INPUT(result, leftOperand, rightOperand);
+
+    if (IS_ZERO(leftOperand)) {
+        NSDecimalCopy(result, rightOperand);
+        return NSCalculationNoError;
+    }
+
+    if (IS_ZERO(rightOperand)) {
+        NSDecimalCopy(result, leftOperand);
+        return NSCalculationNoError;
+    }
+
+    NSDecimal left, right;
+
+    // Copy over decimal for structural manipulation.
+    NSDecimalCopy(&left, leftOperand);
+    NSDecimalCopy(&right, rightOperand);
+
+    // normalize both of them
+    NSCalculationError error = NSDecimalNormalize(&left, &right, roundingMode);
+    if (error != NSCalculationNoError) {
+        return error;
+    }
+
+    // If they are both the same sign, then we can directly add the mantissa
+    if ((IS_NEGATIVE(leftOperand) && IS_NEGATIVE(rightOperand)) || (IS_POSITIVE(leftOperand) && IS_POSITIVE(rightOperand))) {
+        // set the signs
+        NSDecimal* shortest;
+        NSDecimal* longest;
+        unsigned int shortLen;
+        unsigned int longestLen;
+        if (rightOperand->_length < leftOperand->_length) {
+            shortLen = rightOperand->_length;
+            shortest = &right;
+            longestLen = leftOperand->_length;
+            longest = &left;
+            NSDecimalCopy(result, &left);
+
+        } else {
+            shortLen = leftOperand->_length;
+            shortest = &left;
+            longestLen = rightOperand->_length;
+            longest = &right;
+            NSDecimalCopy(result, &right);
+        }
+
+        unsigned short finalResult[NSDecimalMaxSize + 1] = { 0 };
+        unsigned int res = 0;
+
+        for (int index = 0; index < shortLen; ++index) {
+            res += shortest->_mantissa[index] + longest->_mantissa[index];
+            finalResult[index] = res;
+            res = res >> 0x10;
+        }
```

Comment:
i'll standardize. 

---
In reply to: [80589717](https://github.com/Microsoft/WinObjC/pull/1032#discussion_r80589717) [](ancestors = 80589717)